### PR TITLE
Provide positional information when visiting ty, substs and closure_substs in MIR

### PR DESCRIPTION
**Pull request by Nashenas88**
This will enable the region renumbering portion of #43234 (non-lexical lifetimes). @nikomatsakis's current plan [here](https://gist.github.com/nikomatsakis/dfc27b28cd024eb25054b52bb11082f2) shows that we need spans of the original code to create new region variables, e.g. `self.infcx.next_region_var(infer::MiscVariable(span))`. The current visitor impls did not pass positional information (`Location` in some, `Span` and `SourceInfo` for others) for all types. I did not expand this to all visits, just the ones necessary for the above-mentioned plan.

### CODE REVIEW

File: src/librustc/mir/visit.rs [hunk 1]
```diff
@@ -734,6 +737,12 @@ macro_rules! make_mir_visitor {
 make_mir_visitor!(Visitor,);
 make_mir_visitor!(MutVisitor,mut);
 
+pub enum PositionalInfo {
```
Comment (Nashenas88) [inline src/librustc/mir/visit.rs]: This is the one part I wasn't sure about. I could have just passed `Span`s as the third parameter to `visit_ty`, but this leaves the visit implementations a little flexible for future uses. Also, is this the best place for this enum?

**Comment by Nashenas88** [comment]
The one travis failure looks like a network issue. It never even finished the git processes.

File: src/librustc/mir/visit.rs (hunk 1 above)
Comment (arielb1) [inline src/librustc/mir/visit.rs]: I'll just pass a `Span`.

**Comment by arielb1** [comment]
r+ with a `Span` or maybe a `SourceInfo` passed (with `ARGUMENT_VISIBILITY_SCOPE` for the return type). We can always go to passing something more complicated if we need to.

**Comment by Nashenas88** [comment]
Sorry, I made a mistake when I said I could have just passed a `Span`. I had my current work in mind, where the visitor struct held onto a `Mir` struct. I need the `Mir` struct in order to turn the `Location`s into a `Span`. The `Location` struct just holds onto indices for vectors inside the `Mir` struct. Since the traits don't have this, I can't convert the `Location`s to `Span`s without maintaining a copy or the original `Mir` struct and referencing it in all places that call `visit_ty`. I also didn't understand your comment about `ARGUMENT_VISIBILITY_SCOPE`. Where does that come from?

**Comment by carols10cents** [comment]
ping @arielb1, @Nashenas88 has some questions for you!

**Comment by arielb1** [comment]
@Nashenas88 

Then maybe also pass the `SourceInfo` around in all visitor methods? Sounds like it would be useful.

**Comment by Nashenas88** [comment]
@arielb1 I'm not sure I follow. `SourceInfo` data is stored within nested vectors in a `Mir` struct. In order to access those, I need the indices within `Location`. In order to get the `SourceInfo` in all cases, I need both `Mir` and `Location` or just a `Span` as you mentioned above. I could simplify the enum I have to just `Location` and `SourceInfo` given your clarification on `ARGUMENT_VISIBILITY_SCOPE` (thanks!!), though I'm not sure I could simplify this much further without passing around the `Mir` struct. That would pose a challenge given the code is already passing around borrows for the items within the mir, and the only way it exposes access to the vectors is with the top-level `Mir` struct. The vectors inside are private.

**Comment by Nashenas88** [comment]
@arielb1 would it suffice to reduce the enum to `Location` and `SourceInfo`?

**Comment by arielb1** [comment]
Ok

@bors r+

**Pull request merged**